WasatchVCPPLib/WasatchVCPPLib/Spectrometer.cpp

```diff
@@ -192,6 +192,152 @@ bool WasatchVCPP::Spectrometer::setIntegrationTimeMS(unsigned long ms)
     return bytesWritten >= 0;
 }
 
+bool WasatchVCPP::Spectrometer::setLaserPowerPerc(float percent)
+{
+    if (!eeprom.hasLaser)
+    {
+        logger.error("Unable to control laser. EEPROM reports no laser.");
+        return false;
+    }
+    float value = float(max(0, min(100, percent)));
+    laserPowerPerc = value;
+    logger.debug("set_laser_power_perc: range (0, 100), requested %.2f, applying %.2f", percent, value);
+
+	return setLaserPowerPercImmediate(value);
+}
+
+bool WasatchVCPP::Spectrometer::setLaserPowerPercImmediate(float value) {
+
+    value = float(max(0, min(100, value)));
+    // If full power(and allowed), disable modulationand exit
+    if (value >= 100) {
+        uint16_t lsw;
+        uint16_t msw;
+        uint16_t buf[8] = {0, 0, 0, 0, 0, 0, 0, 0};
+
+        if (laserPowerRequireModulation) {
+            logger.debug("100% power requested, yet laser modulation required, so not disabling modulation");
+        }
+        else {
+            logger.debug("Turning off laser modulation (full power)");
+            nextAppliedLaserPower = 100.0;
+            logger.debug("next_applied_laser_power = 100.0");
+            lsw = 0; // disabled;
+            msw = 0;
+            setModEnable(false);
+            return true;
+        }
+    }
+    int period_us = laserPowerHighResolution ? 1000 : 100;
+    int width_us = int((1.0 * value * period_us / 100.0)); // note that value is in range(0, 100) not (0, 1)
+    // pulse width can't be longer than period, or shorter than 1us
+    width_us = max(1, min(width_us, period_us));
+
+    logger.debug("Writing values of period %d and width %d", period_us, width_us);
+    // Change the pulse period.Note that we're not parsing into 40-bit
+    // because this implementation is hard - coded to either 100 or 1000us
+    // (both fitting well within uint16)
+    logger.debug("setting mod period");
+    bool result = setModPeriodus(int(period_us));
+	if (!result) {
+        logger.error("Hardware Failure to send laser mod. pulse period");
+        return false;
+	}
+    getModPeriodus();
+    // Set the pulse width to the 0 - 100 percentage of power
+    logger.debug("setting mod width");
+    result = setModWidthus(int(width_us));
+	if (!result) {
+        logger.error("Hardware Failure to send pulse width");
+        return false;
+	}
+    // Enable modulation
+    logger.debug("enabling mod");
+    result = setModEnable(true);
+	if (!result) {
+		logger.error("Hardware Failure to send laser modulation");
+		return false;
+	}
+    getModEnabled();

```

Comment:
They are for debugging only. I was using them because I was having issues. I thought it better to leave them in case there are plans later to expand Wasatch.VCPP to have more of what's in Wasatch.PY. I'm planning to just add a comment to indicate this. If it's preferred to just remove it then I will.

Reply:
I'd just comment them out so we know it works without them :-)